DIPs/1NNN-WGB.md

```diff
@@ -0,0 +1,155 @@
+pointers escaped a function or not. This implies that a container
+can safely pass a reference to its internals to a function, if that
+function doesn't escape it.
+
+But if the function is also passed another reference to the container,
+that reference can render invalid what the other reference refers to.
+This DIP aims to correct that problem. It's a natural progression after
+DIP25 and DIP1000.
+
+## Contents
+* [Rationale](#rationale)
+* [Description](#description)
+* [Breaking Changes and Deprecations](#breaking-changes-and-deprecations)
+* [Reference](#reference)
+* [Copyright & License](#copyright--license)
+* [Reviews](#reviews)
+
+## Rationale
+
+Containers cannot be memory safe if they cannot control memory safe
+access to their payloads. Containers cannot be efficient if they cannot
+make available direct references to their payloads. Entreating the user
+to just not do certain things is not reliable and does not scale.
+
+The simplest illustration of the problem:
+
+```
+struct S {
+    byte* ptr;
+    ref byte get() { return *ptr; }
+}
+
+void foo(ref S t, ref byte b) {
+    free(t.ptr);    // frees memory referred to by b
+    b = 4;          // corrupt memory access
+}
+
+void test() {
+    S s;
+    s.ptr = cast(byte*)malloc(1);
+    foo(s, s.get());  // (*)
+}
+```
+The same problem using scope pointers:
+```
+struct S {
+    byte* ptr;
+    byte* byte get() { return ptr; }
+}
+
+void foo(scope S* t, scope byte* pb) {
+    free(t.ptr);    // frees memory referred to by pb
+    *pb = 4;        // corrupt memory access
+}
+
+void test() {
+    S s;
+    s.ptr = cast(byte*)malloc(1);
+    foo(&s, s.get());  // (*)
+}
+```
+
+D currently has no defense against this problem. It makes checkably memory safe
+reference counting not possible. (Timon Gehr first pointed this out.)
+
+
+## Description
+
+The solution hinges on the recognition that, in the example, two mutable references
+to the same data are passed to the function `foo()`. Generalizing, whenever there are
+multiple references to the same data, and one of them is mutable, the mutable reference
+can be used to invalidate the data the other references refer to, whether they are
+const or not. Therefore, if more than one reference to the same data is passed to
+a function, they must all be const.
+
+This builds on the foundation built and tested by DIP25 and DIP1000 that track lifetimes
+through function calls and their returns, by adding an additional check on data
+already collected by the compiler semantics.
+
+The checks would only be enforced for @safe code.
+
+### Syntax
+
+No syntax changes. It adds additional semantic checks on existing
+constructs.
+
+### Limitations
+
+This is only checking expressions that are function calls. It does not do interstatement
+checking. It does not check non-scope pointers. It is not a complete borrowing/ownership
+scheme, although it is an important step in that direction. For example, the checking
```

Comment:
Is there a plan or design for a complete borrowing/ownership scheme in D?  If so, it would be nice to read about it in this DIP so we can see where this is all headed.

Reply:
I do have plans, but they are ill-formed.